In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [10]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
lending_data_df = pd.read_csv('lending_data.csv')
# Review the DataFrame
# YOUR CODE HERE!
print(lending_data_df.describe())
lending_data_df.head()


          loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%  

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [12]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_data_df['loan_status']

# Separate the X variable, the features
# YOUR CODE HERE!
x = lending_data_df.drop(columns=['loan_status'])


In [15]:
# Review the y variable Series
# YOUR CODE HERE!
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [16]:
# Review the X variable DataFrame
# YOUR CODE HERE!
x.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [18]:
# Check the balance of our target values
# YOUR CODE HERE!
y_balance = lending_data_df['loan_status'].value_counts()
y_balance

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [20]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [22]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
logistic_regression_model = LogisticRegression(random_state = 1)

# Fit the model using training data
# YOUR CODE HERE!
logistic_regression_model.fit(x_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [23]:
# Make a prediction using the testing data
# YOUR CODE HERE!
y_prediction = logistic_regression_model.predict(x_test)
y_prediction

array([0, 0, 0, ..., 0, 0, 0])

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [31]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
from sklearn.metrics import accuracy_score
balance_accuracy = accuracy_score(y_test, y_prediction)
print(balance_accuracy)

0.9920037831563562


In [32]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_test, y_prediction)
print(confusion)

[[22399   116]
 [   70   676]]


In [33]:
# Print the classification report for the model
# YOUR CODE HERE!
from sklearn.metrics import classification_report
classification = classification_report(y_test, y_prediction)
print(classification)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     22515
           1       0.85      0.91      0.88       746

    accuracy                           0.99     23261
   macro avg       0.93      0.95      0.94     23261
weighted avg       0.99      0.99      0.99     23261



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model appears to perform very well in predicting both the '0' (healthy loan) and '1' (high-risk loan) labels based on the provided metrics, with a very high accuracy (score 0.99 which indicates that it correctly preduct the labels for neary 99% of the test data samples), balanced accuracy score (very high at approximtely 0.992, meaning it is not biased toward the majority class), well-balance precision (22,339 correctly predicted healthy loans and 676 correctly predicted high-risk loans), and recall value for both classes. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [34]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
random_sampler = RandomOverSampler(random_state = 1)
# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
x_train_random, y_train_random = random_sampler.fit_resample(x_train, y_train)

In [36]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
resampled_values = pd.Series(y_train_random).value_counts()
resampled_values

0    52521
1    52521
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [37]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
logistic_re = LogisticRegression(random_state = 1)

# Fit the model using the resampled training data
# YOUR CODE HERE!
logistic_re.fit(x_train_random, y_train_random)
# Make a prediction using the testing data
# YOUR CODE HERE!
y_prediction_re = logistic_re.predict(x_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [38]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
from sklearn.metrics import accuracy_score
balance_accuracy_re = accuracy_score(y_test, y_prediction_re)
print(balance_accuracy_re)

0.9941103134001118


In [39]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
from sklearn.metrics import confusion_matrix
confusion_re = confusion_matrix(y_test, y_prediction_re)
print(confusion_re)

[[22382   133]
 [    4   742]]


In [40]:
# Print the classification report for the model
# YOUR CODE HERE!
from sklearn.metrics import classification_report
classification_re = classification_report(y_test, y_prediction_re)
print(classification_re)


              precision    recall  f1-score   support

           0       1.00      0.99      1.00     22515
           1       0.85      0.99      0.92       746

    accuracy                           0.99     23261
   macro avg       0.92      0.99      0.96     23261
weighted avg       0.99      0.99      0.99     23261



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, fit with oversampled data, predicts both the '0' (healthy loan) and '1' (high-risk loan) labels very effectively. It achives very high accuracy (score 0.994, suggesting the model performs exceptionally well), minimal errors (there are 22,382 correctly predicted healthy loan and 742 correctly predicted high-risk loan, just a few false positives and false negatives), and a well-balanced performance for both classes.